### ISSUE TITLE
bool(True) sent to ODBC with incorrect bit order on little-endian systems

### ISSUE THREAD

**Issue by soundstripe** [mkleehammer/pyodbc#943]
### Environment

- Python: 3.6, 3.9
- pyodbc: 4.0.31 (installed from ibm's yum repo)
- OS: IBM i 7.4
- DB: Db2 for i
- driver: iAccess ODBC driver (1.1.15, ppc64)

### Issue

I've been testing a django app on my Windows dev machine and am in the process of trying to get it to run properly on our IBM i Power9 system. 

Boolean True seems to be sent to the database with the wrong byte order, causing it to be interpreted as a large value instead of 1. Django's BooleanField casts values to bool before sending them to the DB driver, so I cannot send as int without monkeypatching. These are being stored in SMALLINT fields, as Db2 has no built-in Boolean field. Sending numeric data works fine.

I see in `DetectCType` that booleans are treated as SQL_C_BIT and I'm wondering if instead they should be treated as numeric values instead.

I still haven't narrowed down exactly where it is being interpreted/stored as big endian. The whole chain does work, targeting the same DB from Windows/x64. It is possible that the finger needs to be pointed at the ODBC driver here.

```
import pyodbc

conn = pyodbc.connect(
    'DRIVER={IBM i Access ODBC Driver};SYSTEM=x.x.x.x;UID=x;PWD=x'
    # 'DRIVER={IBM i Access ODBC Driver};SYSTEM=192.168.20.101;UID=jamess;PWD=funkymonk'
)
c = conn.cursor()
c.execute(
    'select cast(? as smallint), cast(? as smallint) from sysibm.sysdummy1',
    (1, True)
)
print(list(c.fetchall()))
# [(1, ), (256, )]
```

**Comment by soundstripe** [comment]
Note I may have reversed my usage of big/little endian. I get confused; I just know that somewhere it is getting reversed.

**Comment by v-chojas** [comment]
IBM i is big endian. An ODBC trace will show more clearly what's going on.

**Comment by soundstripe** [comment]
[odbc.log](https://github.com/mkleehammer/pyodbc/files/7015324/odbc.log)

Collected this on the iSeries, running the script as above.

**Comment by gordthompson** [comment]
This is almost certainly a driver issue. For a Python `bool` parameter, pyodbc allocates a 1-byte buffer

https://github.com/mkleehammer/pyodbc/blob/ceeab39598b62d83ad56c9b16a9f40042bb4e766/src/params.cpp#L41-L46

so the byte order of the platform is irrelevant.

Do you have any better luck if you pass `int(True)` instead of `True` as the parameter value?

**Comment by soundstripe** [comment]
Yes. Passing an int works, but Django coerces BooleanField values to bool, so it’s not an easy option (currently using a Middleware to monkey patch this to happen).

Doesn’t bit order often follow byte order? When I use a smallint field I get 256 (2**8), but when I use a larger int field I get a larger number (strangely I think 2**24 but I need to verify). 

I’m going to do some more testing and check with IBM re: the iAccess driver (it is closed source so I can’t look myself).

**Comment by gordthompson** [comment]
> Doesn’t bit order often follow byte order?

My understanding is that the significance of [**bit**-endianness](https://en.wikipedia.org/wiki/Endianness#Bit_endianness) is largely restricted to serial communications (i.e., sending a value bit-by-bit over a serial line like RS-232) and therefore not usually applicable to byte-oriented operations, but then again my recent experience with big-endian systems is extremely limited (read: "none").

The numbers you describe make sense for little-endian encoding being interpreted as big-endian:

2-byte int: 256 = 0x0100  
4-byte int: 16,777,216 = 0x01000000

but again, "embiggening" a 1-byte buffer (as supplied to the DM by pyodbc) into a 2-byte or 4-byte value is something that I would expect the driver to be responsible for.
